lib/IRGen/GenDecl.cpp

```diff
@@ -1697,8 +1697,20 @@ static bool isReadOnlyFunction(SILFunction *f) {
     return false;
 
   auto Eff = f->getEffectsKind();
-  return Eff == EffectsKind::ReadNone ||
-         Eff == EffectsKind::ReadOnly;
+
+  // Swift's readonly does not automatically match LLVM's readonly.
+  // Swift SIL optimizer relies on @effects(readonly) to remove e.g.
+  // dead code remaining from initializers of strings or dictionaries
+  // of variables that are not used. But those initializers are often
+  // not really readonly in terms of LLVM IR. For example, the
+  // Dictionary.init() is marked as @effects(readonly) in Swift, but
+  // it does invoke reference-counting operations.
+  // TODO: Analyze the body of function f and return true if it is
+  // really readonly.
+  if (Eff == EffectsKind::ReadOnly)
+    return false;
+
+  return Eff == EffectsKind::ReadNone;
```

Comment:
I think we should be on the safe side.
If we need a way to specify llvm's readonly in the future we have to change the effects mechanism anyway